File: android/camera/camerax-java/app/src/main/java/com/akhil/cameraxjavademo/MainActivity.java
package com.akhil.cameraxjavademo;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.camera2.interop.Camera2CameraInfo;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraInfo;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageCaptureException;
import androidx.camera.core.ImageProxy;
import androidx.camera.core.Preview;
// import androidx.camera.extensions.HdrImageCaptureExtender;
import androidx.camera.extensions.ExtensionMode;
import androidx.camera.extensions.ExtensionsManager;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.LifecycleOwner;

import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import android.hardware.camera2.CameraCharacteristics;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import com.google.common.util.concurrent.ListenableFuture;

import java.io.File;
import java.lang.reflect.Array;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;
import java.util.Locale;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {
    private final String TAG = "camerax-java-demo";

    private Executor executor = Executors.newSingleThreadExecutor();
    private int REQUEST_CODE_PERMISSIONS = 1001;
    private final String[] REQUIRED_PERMISSIONS = new String[]{"android.permission.CAMERA", "android.permission.WRITE_EXTERNAL_STORAGE"};

    PreviewView mPreviewView;
    ImageView captureImage;

    private int frameCount = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mPreviewView = findViewById(R.id.previewView);
        captureImage = findViewById(R.id.captureImg);

        if(allPermissionsGranted()){
            startCamera(); //start camera if permission has been granted by user
        } else{
            ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS);
        }
    }

    private void startCamera() {

        final ListenableFuture<ProcessCameraProvider> cameraProviderFuture = ProcessCameraProvider.getInstance(this);

        cameraProviderFuture.addListener(new Runnable() {
            @Override
            public void run() {
                try {
                    // copy the following code into the documenation
                    ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                    bindPreview(cameraProvider);

                } catch (ExecutionException | InterruptedException e) {
                    // No errors need to be handled for this Future.
                    // This should never be reached.
                }
            }
        }, ContextCompat.getMainExecutor(this));
    }

    @SuppressLint({"UnsafeOptInUsageError"})
    Boolean isBackCameraLevel3Device(ProcessCameraProvider cameraProvider) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N)
            return Boolean.FALSE;

        int hardwareLevel = -1;

        @SuppressLint("RestrictedApi") List<CameraInfo> cameraInfos = CameraSelector.DEFAULT_BACK_CAMERA
                .filter(cameraProvider.getAvailableCameraInfos());

        if(!cameraInfos.isEmpty()) {
            hardwareLevel = Camera2CameraInfo
                    .from(cameraInfos.get(0))
                    .getCameraCharacteristic(CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL);
        }
        return (hardwareLevel == CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_3);
    }
    void bindPreview(@NonNull ProcessCameraProvider cameraProvider) throws ExecutionException, InterruptedException {

        Log.i(TAG, "Is Level 3 Camera: " + isBackCameraLevel3Device(cameraProvider).toString());
        Preview preview = new Preview.Builder()
                .build();

        CameraSelector cameraSelector = new CameraSelector.Builder()
                .requireLensFacing(CameraSelector.LENS_FACING_BACK)
                .build();

        //Vendor-Extensions (The CameraX extensions dependency in build.gradle)
            /*
                ExtensionMode.FACE_RETOUCH;
                ExtensionMode.BOKEH;
                ExtensionMode.AUTO;
                ExtensionMode.NONE;
                ExtensionMode.HDR;
             */
        int mode = ExtensionMode.NIGHT;
        ExtensionsManager extensionsManager = ExtensionsManager.getInstance(this).get();
        if(extensionsManager != null &&
           extensionsManager.isExtensionAvailable(cameraProvider, cameraSelector, mode))
        {
            cameraSelector = extensionsManager.getExtensionEnabledCameraSelector(cameraProvider, cameraSelector, mode);
            Log.i(TAG, "Extension " + mode + " is supported");
        } else {
            Log.i(TAG, "Extension " + mode + " is not supported");
        }

        ImageAnalysis imageAnalysis = new ImageAnalysis.Builder()
                .build();
        imageAnalysis.setAnalyzer(executor, new ImageAnalysis.Analyzer() {
            int frameCount = 0;
            long timeStamp  = System.currentTimeMillis();
            @Override
            public void analyze(@NonNull ImageProxy imageProxy) {
                frameCount++;
                if (frameCount == 100) {
                    long time = timeStamp;
                    timeStamp = System.currentTimeMillis();
                    Log.i(TAG, "analyzer frameRate = " + (frameCount * 1000.0f / (timeStamp - time)));
                    frameCount = 0;
                }
                imageProxy.close();
            }
        });

        ImageCapture.Builder builder = new ImageCapture.Builder();

        final ImageCapture imageCapture = builder
                .setTargetRotation(this.getWindowManager().getDefaultDisplay().getRotation())
                .build();

        preview.setSurfaceProvider(mPreviewView.getSurfaceProvider());

        Camera camera = null;
        try {
            cameraProvider.bindToLifecycle((LifecycleOwner)this, cameraSelector, preview, imageCapture, imageAnalysis);
        } catch (Exception e) {
            Log.e(TAG, e.getLocalizedMessage());
        }

        captureImage.setOnClickListener(v -> {

            SimpleDateFormat mDateFormat = new SimpleDateFormat("yyyyMMddHHmmss", Locale.US);
            String name = mDateFormat.format(new Date()) + ".jpg";
            File file = new File(getBatchDirectoryName(), name);

            Log.i(this.TAG, "capturing image to: " +  name);
            ImageCapture.OutputFileOptions outputFileOptions = new ImageCapture.OutputFileOptions.Builder(file).build();
            imageCapture.takePicture(outputFileOptions, executor, new ImageCapture.OnImageSavedCallback () {
                @Override
                public void onImageSaved(@NonNull ImageCapture.OutputFileResults outputFileResults) {
                    new Handler(Looper.getMainLooper()).post(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(MainActivity.this, "Image Saved successfully", Toast.LENGTH_SHORT).show();
                        }
                    });
                }
                @Override
                public void onError(@NonNull ImageCaptureException error) {
                    error.printStackTrace();
                }
            });
        });
    }

    public String getBatchDirectoryName() {

        String app_folder_path = "";
        app_folder_path = Environment.getExternalStorageDirectory().toString() + "/images";
        File dir = new File(app_folder_path);
        if (!dir.exists() && !dir.mkdirs()) {

        }

        return app_folder_path;
    }

    private boolean allPermissionsGranted(){

        for(String permission : REQUIRED_PERMISSIONS){
            if(ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED){
                return false;
            }
        }
        return true;
    }
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {

        if(requestCode == REQUEST_CODE_PERMISSIONS){
            if(allPermissionsGranted()){
                startCamera();
            } else{
                Toast.makeText(this, "Permissions not granted by the user.", Toast.LENGTH_SHORT).show();
                this.finish();
            }
        }
    }
}

File: android/camera/camerax/device-selector/src/main/java/com/google/android/preview/MainActivity.kt
package com.google.android.preview

import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.camera.core.ImageCapture
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import java.io.File
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import android.Manifest
import android.annotation.SuppressLint
import android.icu.text.SimpleDateFormat
import android.net.Uri
import android.util.Log
import android.widget.Toast
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import com.google.android.preview.databinding.ActivityMainBinding
import java.nio.ByteBuffer
import java.util.*
import kotlin.system.measureTimeMillis

import androidx.camera.video.*
import java.lang.IllegalArgumentException
import androidx.camera.video.VideoCapture as VideoCaptureX

import com.gfan.android.camerax_lib.CameraSelectorAdapter
typealias LumaListener = (luma: Double) -> Unit


class MainActivity : AppCompatActivity() {
    private lateinit var uiBinding: ActivityMainBinding
    private var imageCapture: ImageCapture? = null
    private lateinit var outputDirectory:File
    private lateinit var cameraExecutor: ExecutorService
    private lateinit var recorder: Recorder
    private lateinit var videoCapturer: VideoCaptureX<Recorder>
    private var activeRecorder: Recording? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        uiBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(uiBinding.root)
        // setContentView(R.layout.activity_main)
        // Request camera permissions
        if (allPermissionsGranted()) {
            startCamera()
        } else {
            ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS,
                REQUEST_CODE_PERMISSIONS)
        }

        // Set up the listener for take photo button
        uiBinding.cameraCaptureButton.setOnClickListener { takePhoto() }
        outputDirectory = getOutputDirectory()
        cameraExecutor = Executors.newSingleThreadExecutor()
    }
    private fun takePhoto() {
        val recordInstance = activeRecorder
        if (recordInstance != null) {
            recordInstance.stop()
            activeRecorder = null

            Log.i("====Capture", "Stopped!")
            return
        }
        // Create time-stamped output file to hold the image
        val photoFile = File(
            outputDirectory,
            SimpleDateFormat(FILENAME_FORMAT, Locale.US
            ).format(System.currentTimeMillis()) + ".3pg")

        // Create output options object which contains file + metadata
        val fileOptions = FileOutputOptions.Builder(photoFile).build()
        activeRecorder = recorder
            .prepareRecording(this, fileOptions)
            .start(cameraExecutor, {
                when (it) {
                    is VideoRecordEvent.Start -> {
                        Log.i(TAG, "Recording Started")
                        uiBinding.cameraCaptureButton.text = "Stop Capture"
                    }
                    is VideoRecordEvent.Status -> Log.i(TAG, "Intermediate Status Report")
                    is VideoRecordEvent.Resume -> Log.i(TAG, "Recording just resumed")
                    is VideoRecordEvent.Pause -> Log.i(TAG, "Recording just paused")
                    is VideoRecordEvent.Finalize -> {
                        val event: VideoRecordEvent.Finalize = it
                        Log.i(
                            TAG,
                            "Recording just finished, status(${event.hasError()}), size(${event.outputResults.outputUri}"
                        )
                        uiBinding.cameraCaptureButton.text= "Start Capture"
                    }
                    else -> Log.e(TAG, "Unknown event type $it")
                }
                Log.i(TAG, "Status: size(${it.recordingStats.numBytesRecorded})," +
                        "time(${it.recordingStats.recordedDurationNanos}/100000000000L)")
            })
    }

    private fun takePhoto2() {
        // Get a stable reference of the modifiable image capture use case
        val imageCapture = imageCapture ?: return

        // Create time-stamped output file to hold the image
        val photoFile = File(
            outputDirectory,
            SimpleDateFormat(FILENAME_FORMAT, Locale.US
            ).format(System.currentTimeMillis()) + ".jpg")

        // Create output options object which contains file + metadata
        val outputOptions = ImageCapture.OutputFileOptions.Builder(photoFile).build()

        // Set up image capture listener, which is triggered after photo has
        // been taken
        imageCapture.takePicture(
            outputOptions, ContextCompat.getMainExecutor(this), object : ImageCapture.OnImageSavedCallback {
                override fun onError(exc: ImageCaptureException) {
                    Log.e(TAG, "Photo capture failed: ${exc.message}", exc)
                }

                override fun onImageSaved(output: ImageCapture.OutputFileResults) {
                    val savedUri = Uri.fromFile(photoFile)
                    val msg = "Photo capture succeeded: $savedUri"
                    Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
                    Log.d(TAG, msg)
                }
            })
    }

    private fun startCamera() {
        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)

        cameraProviderFuture.addListener(Runnable {
            // Used to bind the lifecycle of cameras to the lifecycle owner
            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()

            // Preview
            val preview = Preview.Builder()
                .build()
                .also {
                    it.setSurfaceProvider(uiBinding.viewFinder1.surfaceProvider)
                }

            Log.i(TAG, "====ViewPort: ")

            val preview2 = Preview.Builder()
                .build()
                .also {
                    it.setSurfaceProvider(uiBinding.viewFinder2.surfaceProvider)
                }
            imageCapture = ImageCapture.Builder().build()

            val imageAnalyzer = ImageAnalysis.Builder()
                //.setBackpressureStrategy(STRATEGY_BLOCK_PRODUCER)
                .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
                .build()
                .also {
                    it.setAnalyzer(cameraExecutor,
                        LuminosityAnalyzer {
                            luma ->
                        Log.d(TAG, "Average luminosity: $luma")
                    })
                }
            // Select a Camera with CameraSelectorAdapter
            val cameraSelector:CameraSelector = CameraSelectorAdapter.selectExternalOrBestCamera(cameraProvider) ?: throw IllegalArgumentException()

            // create capture
            recorder = Recorder.Builder().setExecutor(cameraExecutor)
                .build()
            videoCapturer = VideoCaptureX.withOutput<Recorder>(recorder)


            try {
                // Unbind use cases before rebinding
                cameraProvider.unbindAll()

                // Bind use cases to camera
                // cameraProvider.bindToLifecycle(this, cameraSelector, preview, imageCapture, imageAnalyzer, videoCapturer)
                cameraProvider.bindToLifecycle(this, cameraSelector, preview)

            } catch(exc: Exception) {
                Log.e(TAG, "Use case binding failed", exc)
            }

        }, ContextCompat.getMainExecutor(this))
    }
    @SuppressLint("MissingSuperCall")
    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults:
        IntArray) {
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                Toast.makeText(this,
                    "Permissions not granted by the user.",
                    Toast.LENGTH_SHORT).show()
                finish()
            }
        }
    }

    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
        ContextCompat.checkSelfPermission(baseContext, it) == PackageManager
            .PERMISSION_GRANTED
    }

    private fun getOutputDirectory(): File {
        val mediaDir = externalMediaDirs.firstOrNull()?.let {
            File(it, resources.getString(R.string.app_name)).apply { mkdirs() }
        }
        return if (mediaDir != null && mediaDir.exists()) mediaDir else filesDir
    }

    override fun onDestroy() {
        cameraExecutor.shutdown()
        super.onDestroy()
    }

    private class LuminosityAnalyzer(private val listener: LumaListener) : ImageAnalysis.Analyzer {

        private fun ByteBuffer.toByteArray(): ByteArray {
            rewind()    // Rewind the buffer to zero
            val data = ByteArray(remaining())
            get(data)   // Copy the buffer into a byte array
            return data // Return the byte array
        }

        override fun analyze(image: ImageProxy) {
            val timeInMillis = measureTimeMillis {

                val buffer = image.planes[0].buffer
                var copy = ByteBuffer.allocate(buffer.capacity())
                buffer.rewind()
                copy.put(buffer)
                buffer.rewind()
                image.close()

                val data = copy.toByteArray()
                val pixels = data.map { it.toInt() and 0xFF }
                val luma = pixels.average()
                // val luma : Double = .5;
                listener(luma)
            }
            Log.d("Analyzer Function","===Execution Time: $timeInMillis")
        }
    }

    companion object {
        private const val TAG = "CameraXCodelab"
        private const val FILENAME_FORMAT="yyyy-MM-dd-HH-mm-ss-SSS"
        private const val   REQUEST_CODE_PERMISSIONS = 10
        private val REQUIRED_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)
    }
}
